### DISCUSSION TITLE
Colour Correction - Plot and Save are different images.

### DISCUSSION THREAD

**Discussion by xeggbee** [colour-science/colour#1286] [Q&A]
## I`m looking to understant why the image saved is different from the plotted one. Below are the details:

i`d like to save the plotted image!

# Reading the png image:

`RGB = colour.read_image('/Users/moon/Documents/phd/kit_teste4/_BBOX_base_bbox.v2i.yolov7pytorch/png/5-APC_0140.png')

COLOUR_CHECKER_IMAGE_PATHS = glob.glob(
    os.path.join("/Users/moon/Documents/phd/kit_teste4/_BBOX_base_bbox.v2i.yolov7pytorch/png/", "5-APC_0140.png"))

COLOUR_CHECKER_IMAGES = [
    colour.cctf_decoding(RGB)
]
for image in COLOUR_CHECKER_IMAGES:
    colour.plotting.plot_image(colour.cctf_encoding(image));
`

# Identifying the swatches

`SWATCHES = []
for image in COLOUR_CHECKER_IMAGES:
    for colour_checker_swatches_data in detect_colour_checkers_segmentation(
        image, additional_data=True):
        print (colour_checker_swatches_data)
        swatch_colours, colour_checker_image, swatch_masks = (
            colour_checker_swatches_data.values)
        SWATCHES.append(swatch_colours)
        
        # Using the additional data to plot the colour checker and masks.
        masks_i = np.zeros(colour_checker_image.shape)
        for i, mask in enumerate(swatch_masks):
            masks_i[mask[0]:mask[1], mask[2]:mask[3], ...] = 1
        colour.plotting.plot_image(
            colour.cctf_encoding(
                np.clip(colour_checker_image + masks_i * 0.25, 0, 1)));

print (SWATCHES)
D65 = colour.CCS_ILLUMINANTS['CIE 1931 2 Degree Standard Observer']['D65']
REFERENCE_COLOUR_CHECKER = colour.CCS_COLOURCHECKERS[
    'ColorChecker24 - Before November 2014']

REFERENCE_SWATCHES = colour.XYZ_to_RGB(
        colour.xyY_to_XYZ(list(REFERENCE_COLOUR_CHECKER.data.values())),
        REFERENCE_COLOUR_CHECKER.illuminant, D65,
        colour.RGB_COLOURSPACES['sRGB'].matrix_XYZ_to_RGB)`

# now, making the correction and plotting and saving the image:

`for i, swatches in enumerate(SWATCHES):
    swatches_xyY = colour.XYZ_to_xyY(colour.RGB_to_XYZ(
        swatches, D65, D65, colour.RGB_COLOURSPACES['sRGB'].matrix_RGB_to_XYZ))

    colour_checker = colour.characterisation.ColourChecker(
        os.path.basename(COLOUR_CHECKER_IMAGE_PATHS[i]),
        dict(zip(REFERENCE_COLOUR_CHECKER.data.keys(), swatches_xyY)),
        D65)
    
    colour.plotting.plot_multi_colour_checkers(
        [REFERENCE_COLOUR_CHECKER, colour_checker])
    
    #swatches_f = colour.colour_correction(swatches, swatches, REFERENCE_SWATCHES, method = "Vandermonde")

    swatches_f = colour.colour_correction(swatches, swatches, REFERENCE_SWATCHES, method="Finlayson 2015")
    #swatches_f = colour.colour_correction(swatches, swatches, REFERENCE_SWATCHES, method="Cheung 2004")

    swatches_f_xyY = colour.XYZ_to_xyY(colour.RGB_to_XYZ(
        swatches_f, D65, D65, colour.RGB_COLOURSPACES['sRGB'].matrix_RGB_to_XYZ))
    colour_checker = colour.characterisation.ColourChecker(
        '{0} - CC'.format(os.path.basename(COLOUR_CHECKER_IMAGE_PATHS[i])),
        dict(zip(REFERENCE_COLOUR_CHECKER.data.keys(), swatches_f_xyY)),
        D65)
    
    colour.plotting.plot_multi_colour_checkers(
        [REFERENCE_COLOUR_CHECKER, colour_checker])
    print ("printando")
    colour.plotting.plot_image(colour.cctf_encoding(
        colour.colour_correction(
            #COLOUR_CHECKER_IMAGES[i], swatches, REFERENCE_SWATCHES,  method = "Cheung 2004")));
            COLOUR_CHECKER_IMAGES[i], swatches, REFERENCE_SWATCHES,  method = "Finlayson 2015")));
            #COLOUR_CHECKER_IMAGES[i], swatches, REFERENCE_SWATCHES,  method = "Vandermonde")));
    print ("colcoando na variavel")
    corrected_image = colour.colour_correction(
        COLOUR_CHECKER_IMAGES[i], swatches, REFERENCE_SWATCHES, method="Finlayson 2015", degree=1)
    print ("encoding")
    encoded_image = colour.cctf_encoding(corrected_image)
    #colour.plotting.plot_image(encoded_image)
    print ("salvando")
    save_path = '/Users/moon/Documents/phd/kit_teste4/_BBOX_base_bbox.v2i.yolov7pytorch/png/corrected_image.png'
    colour.write_image(encoded_image, save_path)
`

**Comment by thomasmansencal** [comment]
Hello,

How are you displaying the saved image?

Cheers,

Thomas

**Reply by xeggbee** [reply]
@KelSolaar,
thansk for the reply!

Actually,  I was just opening in on Mac OS - Preview. But, If open the saved png:

`
RGB = colour.read_image('/Users/moon/Documents/phd/kit_teste4/_BBOX_base_bbox.v2i.yolov7pytorch/png/corrected_image.png')

COLOUR_CHECKER_IMAGE_PATHS = glob.glob(
    os.path.join("/Users/moon/Documents/phd/kit_teste4/_BBOX_base_bbox.v2i.yolov7pytorch/png/", "corrected_image.png"))

COLOUR_CHECKER_IMAGES = [
    colour.cctf_decoding(RGB)
]
for image in COLOUR_CHECKER_IMAGES:
    colour.plotting.plot_image(colour.cctf_encoding(image));
`




## i also saved the corrected as tiff, using tifffile

`RGB = colour.read_image('/Users/moon/Documents/phd/kit_teste4/_BBOX_base_bbox.v2i.yolov7pytorch/png/corrected_image.tiff')

COLOUR_CHECKER_IMAGE_PATHS = glob.glob(
    os.path.join("/Users/moon/Documents/phd/kit_teste4/_BBOX_base_bbox.v2i.yolov7pytorch/png/", "corrected_image.tiff"))

COLOUR_CHECKER_IMAGES = [
    colour.cctf_decoding(RGB)
]
for image in COLOUR_CHECKER_IMAGES:
    colour.plotting.plot_image(colour.cctf_encoding(image));
`



## the same corrected image, but on macos preview:


## Gimp, same corrected image

**Reply by thomasmansencal** [reply]
So, it is almost certainly double-encoding: When you read a PNG file, irrespective of whether it is 8-bit or 16-bit, it will almost certainly have been encoded non-linearly to avoid banding. Now, when you bring it in Colour, the default is to convert it to Float64 but it stays non-linearly encoded. Depending on what you are intending to do with the image, you will want to decode it after reading, e.g., use `colour.cctf_decoding` function.

Now in your previous code, you read the image but do not decode it and before displaying it, you are re-encoding it, thus encoding two times.

Hope this makes sense.
